File: backend/service/quiz_exam_service/app/schemas/quiz_submission.py
from pydantic import BaseModel
from uuid import UUID
from typing import List, Optional
from app.models.enum import SubmissionStatus , QuestionType
from datetime import datetime

class QuizSubmissionCreate(BaseModel):
    quiz_id: UUID
    user_id: UUID
    attempt_number: Optional[int] = None

class QuizSubmissionUpdate(BaseModel):
    status: SubmissionStatus | None = None
    submitted_at: Optional[datetime] = None
    # Chấm điểm chéo
    peer_avg_score: Optional[float] = None
    is_discrepant: bool | None = None
    completed_review_count: int | None = None 
    
    total_score: Optional[float] = None 
    is_passed: Optional[bool] = None 
    grader_id: Optional[UUID] = None

class SubmissionStatusOption(BaseModel):
    option_id: UUID
    option_text: str
    is_correct: Optional[bool] = None  # Chỉ có giá trị khi bài đã được chấm

class SubmissionStatusDetail(BaseModel):
    detail_id: UUID
    question_id: UUID
    question_title: str
    question_type: QuestionType
    video_trigger_seconds: Optional[int] = None
    body_content: Optional[str]
    max_points: Optional[float]
    options: List[SubmissionStatusOption]

    selected_option_id: Optional[UUID] = None
    essay_answer_text: Optional[str] = None
    graph_json_data: Optional[str] = None
    graph_image_url: Optional[str] = None

    # Chỉ trả khi bài đã được chấm
    score_earned: Optional[float] = None
    teacher_feedback: Optional[str] = None

class QuizSubmissionStatusResponse(BaseModel):
    submission_id: UUID
    quiz_id: UUID
    status: SubmissionStatus
    attempt_number: int
    started_at: datetime
    total_score: Optional[float] = None
    is_passed: Optional[bool] = None
    questions: List[SubmissionStatusDetail]

File: frontend/src/types/quiz-submission.ts
import { SubmissionStatus } from "./statuses";

export enum QuestionType {
    MULTIPLE_CHOICE = "MULTIPLE_CHOICE",
    ESSAY = "ESSAY",
    TRUE_FALSE = "TRUE_FALSE",
}

export interface QuestionOption {
    option_id: string;
    option_text: string;
}

export interface QuizQuestion {
    detail_id: string;
    question_id: string;
    question_title: string;
    question_type: QuestionType;
    body_content: string | null;
    max_points: number;
    options: QuestionOption[];
    // 🆕 Mốc giây trong video mà câu hỏi này được kích hoạt (null nếu quiz không phải IN_VIDEO)
    video_trigger_seconds?: number | null;
    // 🆕 Dùng khi resume submission IN_PROGRESS đang dang dở (vd load lại trang giữa video)
    selected_option_id?: string | null;
    // 🆕 true/false nếu câu đã được chấm, null nếu chưa trả lời hoặc là ESSAY
    is_answered_correct?: boolean | null;
}

export interface QuizTakeResponse {
    submission_id: string;
    quiz_id: string;
    title: string;
    quiz_type: string;
    attempt_number: number;
    questions: QuizQuestion[];
}

export interface SubmissionDetailUpdatePayload {
    selected_option_id?: string | null;
    essay_answer_text?: string | null;
}

export interface SubmissionDetailUpdateResponse {
    message: string;
    detail_id: string;
    selected_option_id: string | null;
    essay_answer_text: string | null;
}

export interface QuizSubmitResponse {
    message: string;
    submission_id: string;
    status: string;
    is_passed: boolean;
    total_score: number;
    next_lesson_unlocked: boolean;
}

export interface SubmissionStatusOption {
    option_id: string;
    option_text: string;
    is_correct?: boolean | null; // Chỉ có khi status = GRADED
}

export interface SubmissionStatusDetail {
    detail_id: string;
    question_id: string;
    question_title: string;
    question_type: QuestionType;
    video_trigger_seconds?: number | null; // 🆕
    body_content?: string | null;
    max_points?: number | null;
    options: SubmissionStatusOption[];
    selected_option_id?: string | null;
    essay_answer_text?: string | null;
    graph_json_data?: string | null;
    graph_image_url?: string | null;
    score_earned?: number | null;     // Chỉ có khi status = GRADED
    teacher_feedback?: string | null; // Chỉ có khi status = GRADED
}

export interface QuizSubmissionStatusResponse {
    submission_id: string;
    quiz_id: string;
    title?: string | null;
    quiz_type?: string | null;
    status: SubmissionStatus;
    attempt_number: number;
    started_at: string;
    total_score?: number | null;
    is_passed?: boolean | null;
    questions: SubmissionStatusDetail[];
}

export interface QuizStatusActionResult {
    success: boolean;
    data?: QuizSubmissionStatusResponse | null; // null nghĩa là chưa từng làm bài
    error?: string;
}